File: 2OLIDTS-WilliamSelvas-09/2OLIDTS-WilliamSelvas-09.cpp
// 2OLIDTS-WilliamSelvas-09.cpp : Este archivo contiene la función "main". La ejecución del programa comienza y termina ahí.
//

#include <iostream>
#define MaxTamC 10

typedef int TipoDato;

using namespace std;

int main(){

	int frente = 0, final = 0;
	TipoDato A[MaxTamC];
	int contador = 0;

	// Inicializar la cola 
	frente = 0;
	final = 0;

	cout << " Desea agregar elementos a la cola? (s/n): ";
	char respuesta;
	cin >> respuesta;

	while ((respuesta == 's' || respuesta == 'S') && contador < 10) {
		if ((final + 1) % MaxTamC == frente) {
			cout << "Desbordamiento de la cola" << endl;
			return 1;
		}

		cout << "Ingrese un elemento para la cola: ";
		TipoDato elemento;
		cin >> elemento;
		
		final = (final + 1) % MaxTamC;
		A[final] = elemento;

		contador++;
		cout << "Elemento " << contador << " agregado a la cola: " << elemento << endl;

		if (contador < 10) {
			cout << "Desea agregar mas elementos a la cola? (s/n): ";
			cin >> respuesta;
		}
	}

	// Validar si la cola está vacía
	if (frente == final) {
		cout << "La cola esta vacía. " << endl;
		return 1;
	}

	// Obtener el primer elemento de la cola 
	TipoDato primerElemento = A[(frente + 1) & MaxTamC];
	cout << "El primer elemento de la cola es: " << primerElemento << endl;

	//Elimina un elemento de la cola 
	frente = (frente + 1) % MaxTamC;

	//Imprimir elemento de la cola en el orden en que fueron ingresados
	cout << "Elementos  de la cola en el orden de ingreso: " << endl;
	for (int i = frente + 1; i <= final; i = (i + 1) % MaxTamC) {
		cout << A[i] << ""; 
	}

	cout << endl;
	
	return 0;
}

File: 2OLIDTS-WilliamSelvas-09/2OLIDTS-WilliamSelvas-09.py
MAX_TAM = 10

def ingresar_valores(frente, final, cola, contador):
    if (final + 1) % MAX_TAM == frente:
        print("Desbordamiento de la cola")
        return frente, final, cola, contador
    
    elemento = int(input("Ingrese un elemento para la cola: "))
    final = (final + 1) % MAX_TAM
    cola[final] = elemento
    contador += 1
    print("Elemento", contador, "agregado a la cola:", elemento)
    return frente, final, cola, contador

def eliminar_elemento(frente, final, contador):
    if frente == final:
        print("La cola esta vacia, no se puede eliminar")
        return frente, contador
    
    frente = (frente + 1) % MAX_TAM
    contador -= 1
    print("Elemento eliminado de la cola.")
    return frente, contador

def imprimir(frente, final, cola):
    print("Elementos de la cola en el orden de ingreso:")
    i = (frente + 1) % MAX_TAM
    while i != (final + 1) % MAX_TAM:
        print(" " + str(cola[i]))
        i = (i + 1) % MAX_TAM
    print()

def main():
    cola = [0] * MAX_TAM
    contador = 0
    frente = 0
    final = 0
    
    respuesta = input("Desea agregar elementos a la cola? (s/n): ").lower()
    while respuesta == "s" and contador < 10:
        frente, final, cola, contador = ingresar_valores(frente, final, cola, contador)
        if contador < 10:
            respuesta = input("Desea agregar mas elementos a la cola? (s/n): ").lower()
            
    if frente == final:
        print("La cola esta vacia.")
        return
    
    primer_elemento = cola[(frente + 1) % MAX_TAM]
    print("El primer elemento de la cola es:", primer_elemento)
    frente, contador = eliminar_elemento(frente, final, contador)
    imprimir(frente, final, cola)

if __name__ == "__main__":
    main()
